pallets/poa/src/lib.rs

```diff
@@ -0,0 +1,710 @@
+    }
+);
+
+// The pallet's errors
+decl_error! {
+    /// Errors for the module.
+    pub enum Error for Module<T: Trait> {
+        MaxValidators,
+        AlreadyActiveValidator,
+        AlreadyQueuedForAddition,
+        AlreadyQueuedForRemoval,
+        NeedAtLeast1Validator,
+        SwapOutFailed,
+        SwapInFailed,
+    }
+}
+
+decl_module! {
+    /// The module declaration.
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        // Do maximum of the work in functions to `add_validator` or `remove_validator` and minimize the work in
+        // `should_end_session` since that it called more frequently.
+
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Add a new validator to active validator set unless already a validator and the total number
+        /// of validators don't exceed the max allowed count. The validator is considered for adding at
+        /// the end of this epoch unless `short_circuit` is true. If a validator is already added to the queue
+        /// an error will be thrown unless `short_circuit` is true, in which case it swallows the error.
+        // Weight can be 0 as its called by Master. TODO: Use signed extension to make it free
+        #[weight = 0]
+        pub fn add_validator(origin, validator_id: T::AccountId, short_circuit: bool) -> dispatch::DispatchResult {
+            // TODO: Check the origin is Master
+            ensure_root(origin)?;
+
+            Self::add_validator_(validator_id, short_circuit)
+        }
+
+        /// Remove the given validator from active validator set and the queued validators at the end
+        /// of epoch unless `short_circuit` is true. If validator is already queued for removal, an error
+        /// will be thrown unless `short_circuit` is true, in which case it swallows the error.
+        /// It will not remove the validator if the removal will cause the active validator set to
+        /// be empty even after considering the queued validators.
+        // Weight can be 0 as its called by Master. TODO: Use signed extension to make it free
+        #[weight = 0]
+        pub fn remove_validator(origin, validator_id: T::AccountId, short_circuit: bool) -> dispatch::DispatchResult {
+            // TODO: Check the origin is Master
+            ensure_root(origin)?;
+            Self::remove_validator_(validator_id, short_circuit)
+        }
+
+        /// Replace an active validator (`old_validator_id`) with a new validator (`new_validator_id`)
+        /// without waiting for epoch to end. Throws error if `old_validator_id` is not active or
+        /// `new_validator_id` is already active. Also useful when a validator wants to rotate his account.
+        // Weight can be 0 as its called by Master. TODO: Use signed extension to make it free
+        #[weight = 0]
+        pub fn swap_validator(origin, old_validator_id: T::AccountId, new_validator_id: T::AccountId) -> dispatch::DispatchResult {
```

Comment:
`add_validator` and `remove_validator` both preserve fairness, even when `short_circuit == true`, but `swap_validator` does not preserve fairness because it causes the session to end as soon as possible. Why is this?

Reply:
Swap is supposed to be used in urgent scenarios (like key rotation as the code comment describes)